lsp/src/main/java/org/aya/lsp/options/RenderOptions.java

```diff
@@ -0,0 +1,216 @@
+
+    boolean modified = false;
+
+    for (var entry : override.entrySet()) {
+      var key = entry.getKey();
+      var value = entry.getValue();
+
+      var isKeyValid = key != null && Seq.from(AyaColorScheme.Key.values()).view()
+        .map(AyaColorScheme.Key::key)
+        .contains(key);
+      var isValueValid = value != null;   // I don't like to inline this code, sorry!
+
+      if (!isKeyValid) return Result.err("Invalid key: " + key);
+      if (!isValueValid) return Result.err("Invalid color: null");
+
+      var colorResult = parseColor(value);
+      if (colorResult.isErr()) return Result.err("Invalid color: " + value);
+
+      var color = colorResult.get();
+      var oldValue = newColorScheme.put(key, color);
+
+      if (oldValue.isDefined() && ! oldValue.get().equals(color)) {
+        modified = true;
+      }
+    }
+
+    return Result.ok(
+      modified ? new AyaColorScheme(newColorScheme) : base);
+  }
+
+  /**
+   * Parsing the color code into an integer. I didn't find any function do this.
+   *
+   * @param color the color string, which should be a hexadecimal number string in 6 numbers long, optional prefix "#" and "0x" are allowed.
+   */
+  public static Result<Integer, String> parseColor(String color) {
+    var colorCode = color;
+    if (color.charAt(0) == '#') colorCode = colorCode.substring(1);
+    if (colorCode.startsWith("0x")) colorCode = colorCode.substring(2);
+
+    if (colorCode.length() != 6) return Result.err("The color code is too long or too short!");
+
+    try {
+      var result = Integer.parseInt(colorCode, 16);
+
+      return Result.ok(result);
+    } catch (NumberFormatException e) {
+      return Result.err(e.getMessage());
+    }
+  }
+
+  /// endregion
+
+  @Contract(pure = true)
+  public @NotNull RenderOptions update(@NotNull RenderOptions other) {
+    var colorScheme = other.colorScheme();
+    var styleFamily = other.styleFamily();
+
+    if (colorScheme.isEmpty()) colorScheme = this.colorScheme();
+    if (styleFamily.isEmpty()) styleFamily = this.styleFamily();
+
+    return new RenderOptions(colorScheme, styleFamily);
+  }
+
+  /// region Render
+
+  /**
+   * Choose the most colorful target which the client supports.
+   * Note that the client should support `plaintext` at least.
+   *
+   * @param capabilities the client capabilities, `plaintext` is returned if the capabilities is empty.
+   * @throws IllegalArgumentException is thrown if the capabilities is not empty and invalid.
+   */
+  public static @NotNull RenderTarget renderTargetFromCapabilities(@NotNull List<String> capabilities)
+    throws IllegalArgumentException {
+    if (capabilities.isEmpty()) return RenderTarget.Debug;
+    if (capabilities.contains(MarkupKind.Markdown)) return RenderTarget.HTML;
+    if (capabilities.contains(MarkupKind.PlainText)) return RenderTarget.Debug;
+
+    throw new IllegalArgumentException("Unsupported capabilities: " + capabilities);
```

Comment:
Should not throw here, instead, return `RenderTarget.Debug` because the client should at least support plain text format.

If you want to tell the client "I don't understand", you should do it in LSP initialization (even in that case you should only raise a warning) where the capabilities are received.